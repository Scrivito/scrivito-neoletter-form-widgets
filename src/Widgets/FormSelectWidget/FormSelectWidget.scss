@keyframes fadeIn {
  from {
    opacity: 0;
  }
  to {
    opacity: 1;
  }
}

.scrivito-neoletter-form-widgets .select-container {
  .select-title {
    display: flex;
    align-items: baseline;
  }
  .select-label {
    display: flex;
    align-items: center;
    span {
      margin-left: 0.4em;
    }
    .form-check-input {
      margin-top: 0;
    }
  }
  .dropdown-select {
    width: 100%;
  }
  .inline {
    display: flex;
    flex-direction: row;
    flex-wrap: nowrap;
    & > * {
      padding-right: 1em;
    }
  }
  .linear-scale-container {
    display: flex;
    .lower-scale-label {
      top: 20px;
      position: relative;
      margin-right: 1.5em;
      align-self: flex-start;
    }
    .upper-scale-label {
      top: 5px;
      position: relative;
      margin-left: 1.5em;
      align-self: flex-end;
    }
    .linear-scale {
      flex-direction: column-reverse;
      width: auto;
      span {
        margin-left: 0;
      }
    }
  }
  .text-end {
    margin-top: 0.5em;
    .reset-label {
      width: fit-content;
      padding: 0.1em 0.6em;
      display: inline-block;
      border-radius: 2px;
      cursor: pointer;
      span {
        color: #5f6368;
      }
      &:hover {
        background-color: #f9f9f9;
      }
    }
    &.fade-in {
      animation: fadeIn 0.2s ease-in-out forwards;
    }
  }
}
