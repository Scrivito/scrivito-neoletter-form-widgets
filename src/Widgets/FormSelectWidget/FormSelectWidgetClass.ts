import * as Scrivito from "scrivito";

export const FormSelectWidget = Scrivito.provideWidgetClass(
  "FormSelectWidget",
  {
    attributes: {
      selectionType: [
        "enum",
        {
          values: ["radio", "dropdown", "multi", "linear-scale"]
        }
      ],
      title: "html",
      items: "stringlist",
      customFieldName: "string",
      required: "boolean",
      helpText: "html",
      linearScaleLowerLimit: ["enum", { values: ["0", "1"] }],
      linearScaleUpperLimit: [
        "enum",
        { values: ["2", "3", "4", "5", "6", "7", "8", "9", "10"] }
      ],
      linearScaleLowerLabel: "string",
      linearScaleUpperLabel: "string",
      clearSelectionButtonText: "string",
      inlineView: "boolean",
      useFloatingLabel: "boolean",
      navigateOnClick: "boolean",
      showClearSelectionButton: "boolean"
    }
  }
);
