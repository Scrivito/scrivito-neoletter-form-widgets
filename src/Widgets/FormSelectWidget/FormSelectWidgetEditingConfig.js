import * as Scrivito from "scrivito";
import { customFieldNameValidation } from "../FormContainerWidget/utils/validations/customFieldNameValidation";
import { insideFormContainerValidation } from "../FormContainerWidget/utils/validations/insideFormContainerValidation";
import formSelectWidgetIcon from "../../assets/images/form_widget_select.svg";

Scrivito.provideEditingConfig("FormSelectWidget", {
  title: "Form Select",
  thumbnail: formSelectWidgetIcon,
  attributes: {
    selectionType: {
      title: "Input type",
      values: [
        { value: "radio", title: "Radio buttons" },
        { value: "dropdown", title: "Dropdown" },
        { value: "multi", title: "Checkboxes" },
      ],
    },
    items: {
      title: "Items",
    },
    title: { title: "Label" },
    customFieldName: { title: "Field name" },
    required: { title: "Mandatory" },
    helpText: { title: "Help text" },
  },
  properties: (obj) => [
    "selectionType",
    "title",
    "items",
    "customFieldName",
    ["required", { enabled: obj.get("selectionType") !== "multi" }],
    "helpText",
  ],
  initialContent: {
    selectionType: "radio",
    title: "Please choose",
    items: ["Yes", "No"],
    customFieldName: "custom_",
  },
  validations: [
    insideFormContainerValidation,
    [
      "items",

      (items) => {
        if (items.length < 2) {
          return {
            message: "The widget must include at least two items.",
            severity: "error",
          };
        }
      },
    ],
    customFieldNameValidation,
  ],
});
