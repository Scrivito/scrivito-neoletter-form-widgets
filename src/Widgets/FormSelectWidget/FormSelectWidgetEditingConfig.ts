import * as Scrivito from "scrivito";
import { customFieldNameValidation } from "../FormStepContainerWidget/utils/validations/customFieldNameValidation";
import { insideFormContainerValidation } from "../FormStepContainerWidget/utils/validations/insideFormContainerValidation";
import formSelectWidgetIcon from "../../assets/images/form_widget_select.svg";

Scrivito.provideEditingConfig("FormSelectWidget", {
  title: "Neoletter Form Select",
  thumbnail: formSelectWidgetIcon,
  attributes: {
    selectionType: {
      title: "Input type",
      values: [
        { value: "radio", title: "Radio buttons" },
        { value: "dropdown", title: "Dropdown" },
        { value: "multi", title: "Checkboxes" },
        { value: "linear-scale", title: "Linear scale" }
      ]
    },
    items: {
      title: "Items"
    },
    title: { title: "Label" },
    customFieldName: { title: "Field name" },
    required: { title: "Mandatory" },
    helpText: { title: "Help text" },
    linearScaleLowerLimit: { title: "Lower scale limit" },
    linearScaleUpperLimit: { title: "Upper scale limit" },
    linearScaleLowerLabel: { title: "Optional label for lower scale limit" },
    linearScaleUpperLabel: { title: "Optional label for upper scale limit" },
    clearSelectionText: { title: "Clear selection text" },
    inlineView: {
      title: "Arrange items horizontally",
      description: "When enabled, all items will be displayed in a single row."
    }
  },
  properties: widget => {
    return getProperties(widget);
  },
  initialContent: {
    selectionType: "radio",
    title: "Please choose",
    items: ["Yes", "No"],
    customFieldName: "custom_",
    linearScaleLowerLimit: "1",
    linearScaleUpperLimit: "5",
    clearSelectionText: "Clear selection",
    inlineView: false
  },
  validations: [
    insideFormContainerValidation,
    [
      "items",

      (items: string[]) => {
        if (items.length < 2) {
          return {
            message: "The widget must include at least two items.",
            severity: "error"
          };
        }
      }
    ],
    customFieldNameValidation
  ]
});

// eslint-disable-next-line @typescript-eslint/no-explicit-any
function getProperties(widget: Scrivito.Obj): any[] {
  const type = widget.get("selectionType");
  const props = [
    "selectionType",
    "title",
    "customFieldName",
    ["required", { enabled: type !== "multi" }],
    "helpText"
  ];
  // show/hide inlineView
  if (type == "radio" || type == "multi") {
    props.splice(3, 0, "inlineView");
  }
  // show/hide items
  if (type != "linear-scale") {
    props.splice(2, 0, "items");
  } else {
    // show/hide linear scale props
    props.splice(
      2,
      0,
      "linearScaleLowerLimit",
      "linearScaleUpperLimit",
      "linearScaleLowerLabel",
      "linearScaleUpperLabel"
    );
  }
  // show/hide clearSelectionText
  if (!widget.get("required") && (type == "linear-scale" || type == "radio")) {
    props.splice(props.length - 1, 0, "clearSelectionText");
  }
  return props;
}
