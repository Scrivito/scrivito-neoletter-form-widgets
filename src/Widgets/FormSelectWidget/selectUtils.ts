import { Widget } from "scrivito";

export const isAlignmentEnabled = (widget: Widget) => {
	const inlineViewEnabled = widget.get("inlineView") as boolean;
	const floatingLabelEnabled = widget.get("useFloatingLabel") as boolean;
	const type = widget.get("selectionType") as string;
	if (type == "dropdown") {
		return !floatingLabelEnabled;
	}
	if (type == "radio" || type == "multi") {
		return inlineViewEnabled;
	}
	return true;
}
