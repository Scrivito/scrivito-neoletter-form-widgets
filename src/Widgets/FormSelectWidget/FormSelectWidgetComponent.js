import * as React from "react";
import * as Scrivito from "scrivito";
import { InPlaceEditingPlaceholder } from "../../Components/InPlaceEditingPlaceholder";
import "./FormSelectWidget.scss";
import { getFieldName } from "../FormContainerWidget/utils/getFieldName";
import { Select } from "../FormContainerWidget/components/SelectComponent";
import { Mandatory } from "../FormContainerWidget/components/MandatoryComponent";
import { HelpText } from "../FormContainerWidget/components/HelpTextComponent";
import { Dropdown } from "../FormContainerWidget/components/SelectDropdownComponent";

Scrivito.provideComponent("FormSelectWidget", ({ widget }) => {
  const items = widget.get("items");
  const isMultiSelect = widget.get("selectionType") == "multi";
  const isDropdown = widget.get("selectionType") == "dropdown";
  if (!items.length) {
    return (
      <InPlaceEditingPlaceholder center>
        Add items in the widget properties.
      </InPlaceEditingPlaceholder>
    );
  }

  return (
    <div className="select-container mb-3">
      <div className="select-title">
        <span className="text-super"> {widget.get("title")} </span>
        {!isMultiSelect && widget.get("required") && <Mandatory />}
        {widget.get("helpText") && <HelpText widget={widget} />}
      </div>
      {isDropdown ? (
        <Dropdown
          id={widget.id()}
          name={getFieldName(widget)}
          options={items}
          required={widget.get("required")}
        />
      ) : (
        <Select
          isMultiSelect={isMultiSelect}
          items={widget.get("items")}
          required={widget.get("required")}
          name={getFieldName(widget)}
        />
      )}
    </div>
  );
});
