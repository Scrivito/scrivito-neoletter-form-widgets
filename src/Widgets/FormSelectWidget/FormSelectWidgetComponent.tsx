import * as React from "react";
import * as Scrivito from "scrivito";
import { InPlaceEditingPlaceholder } from "../../Components/InPlaceEditingPlaceholder";
import { getFieldName } from "../FormStepContainerWidget/utils/getFieldName";
import { Select } from "../FormStepContainerWidget/components/SelectComponent";
import { Mandatory } from "../FormStepContainerWidget/components/MandatoryComponent";
import { HelpText } from "../FormStepContainerWidget/components/HelpTextComponent";
import { Dropdown } from "../FormStepContainerWidget/components/SelectDropdownComponent";
import { FormSelectWidget } from "./FormSelectWidgetClass";
import { ResetInputs } from "../FormStepContainerWidget/components/ResetInputsComponent";
import "./FormSelectWidget.scss";

Scrivito.provideComponent(FormSelectWidget, ({ widget }) => {
  const ref = React.useRef<HTMLDivElement>(null);
  const [selected, setSelected] = React.useState(false);
  const items = widget.get("items");
  const isMultiSelect = widget.get("selectionType") == "multi";
  const isDropdown = widget.get("selectionType") == "dropdown";
  if (!items.length && widget.get("selectionType") != "linear-scale") {
    return (
      <InPlaceEditingPlaceholder center>
        Add items in the widget properties.
      </InPlaceEditingPlaceholder>
    );
  }

  return (
    <div className="select-container mb-3" ref={ref}>
      <div className="select-title">
        <span className="text-super"> {widget.get("title")} </span>
        {!isMultiSelect && widget.get("required") && <Mandatory />}
        {widget.get("helpText") && <HelpText widget={widget} />}
      </div>
      {isDropdown ? (
        <Dropdown
          id={widget.id()}
          name={getFieldName(widget)}
          options={items}
          required={widget.get("required")}
        />
      ) : (
        <Select
          isMultiSelect={isMultiSelect}
          required={widget.get("required")}
          widget={widget}
          name={getFieldName(widget)}
          onChange={() => setSelected(true)}
        />
      )}
      {showReset() && (
        <ResetInputs
          setSelectedCallback={setSelected}
          text={widget.get("clearSelectionText")}
          parentRef={ref}
        />
      )}
    </div>
  );
  function showReset(): boolean {
    return (
      selected &&
      !widget.get("required") &&
      (widget.get("selectionType") == "radio" ||
        widget.get("selectionType") == "linear-scale")
    );
  }
});
