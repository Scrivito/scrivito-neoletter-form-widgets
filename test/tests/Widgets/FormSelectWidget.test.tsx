import * as Scrivito from "scrivito";
import { screen, fireEvent } from "@testing-library/react";
import { FormSelectWidget } from "../../../src/Widgets/FormSelectWidget/FormSelectWidgetClass";
import "../../../src/Widgets/FormSelectWidget/FormSelectWidgetComponent";
import PageRenderer from "../../helpers/pageRenderer";

Scrivito.configure({ tenant: "inMemory" });

const pageRenderer = new PageRenderer();

const widgetProps = {
  title: "Test Select Title",
  helpText: "Sample help text",
  selectionType: "dropdown",
  required: true,
  items: ["Item 1", "Item 2", "Item 3"],
  clearSelectionText: "Clear Selection",
  linearScaleLowerLimit: "1",
  linearScaleUpperLimit: "5",
  linearScaleLowerLabel: "low",
  linearScaleUpperLabel: "high",
  inlineView: false
};

describe("FormSelectWidget", () => {
  it("renders FormSelectWidget as a dropdown", () => {
    pageRenderer.render({
      body: [new FormSelectWidget(widgetProps)]
    });

    const title = screen.getByText(widgetProps.title);
    const selectTitle = document.querySelector(".select-title");
    const dropdown = document.querySelector(".dropdown-select");

    expect(title).toBeInTheDocument();
    expect(selectTitle).toBeInTheDocument();
    expect(dropdown).toBeInTheDocument();
    expect(dropdown).toHaveAttribute("required");
  });

  it("renders FormSelectWidget as a select list", () => {
    const selectProps = { ...widgetProps, selectionType: "radio" };

    pageRenderer.render({
      body: [new FormSelectWidget(selectProps)]
    });

    const title = screen.getByText(selectProps.title);
    const selectTitle = document.querySelector(".select-title");
    const radioInputs = document.querySelectorAll(".form-check-input");

    expect(title).toBeInTheDocument();
    expect(selectTitle).toBeInTheDocument();
    expect(radioInputs.length).toBe(selectProps.items.length);

    radioInputs.forEach((radioInput) => {
      expect(radioInput).toHaveAttribute("required");
    });
  });

  it("handles selection change in the select list", () => {
    const selectProps = { ...widgetProps, selectionType: "radio" };

    pageRenderer.render({
      body: [new FormSelectWidget(selectProps)]
    });

    const radioInput = document.querySelector(
      '[type="radio"][value="Item 2"]'
    )!;
    fireEvent.change(radioInput, { target: { checked: true } });

    expect(radioInput).toBeInTheDocument();
    expect(radioInput).toBeChecked();
  });

  it("renders FormSelectWidget with checkboxes and inlineView", () => {
    const selectProps = {
      ...widgetProps,
      inlineView: true,
      selectionType: "multi",
      required: false
    };

    pageRenderer.render({
      body: [new FormSelectWidget(selectProps)]
    });

    const title = screen.getByText(selectProps.title);
    const selectTitle = document.querySelector(".select-title");
    const checkboxes = document.querySelectorAll(".form-check-input"); // Update with your actual checkbox input class name
    const selectContainer = document.querySelector(".inline")!; // Update with your actual class name for inlineView

    expect(title).toBeInTheDocument();
    expect(selectTitle).toBeInTheDocument();
    expect(checkboxes.length).toBe(selectProps.items.length); // Check the number of checkboxes

    checkboxes.forEach((checkbox) => {
      expect(checkbox).toHaveAttribute("type", "checkbox");
      expect(checkbox).not.toHaveAttribute("required"); // Check if the checkboxes don't have 'required' attribute
    });

    expect(selectContainer).toBeInTheDocument();
    expect(selectContainer.classList).toContain("inline");
  });

  it("renders FormSelectWidget with linear-scale selection", () => {
    const selectProps = {
      ...widgetProps,
      selectionType: "linear-scale"
    };

    pageRenderer.render({
      body: [new FormSelectWidget(selectProps)]
    });

    const linearScaleContainer = document.querySelector(
      ".linear-scale-container"
    );
    const lowerScaleLabel = screen.getByText(selectProps.linearScaleLowerLabel);
    const upperScaleLabel = screen.getByText(selectProps.linearScaleUpperLabel);
    const inputRange = document.querySelectorAll(".form-check-input");

    expect(linearScaleContainer).toBeInTheDocument();
    expect(lowerScaleLabel).toBeInTheDocument();
    expect(upperScaleLabel).toBeInTheDocument();
    expect(inputRange).toHaveLength(5);

    inputRange.forEach((input) => {
      expect(input).toHaveAttribute("required");
      expect(input).toHaveAttribute("type", "radio");
    });
  });

  it("renders correctly", () => {
    const tree = pageRenderer.getAsJSON({
      body: [new FormSelectWidget(widgetProps)]
    });
    expect(tree).toMatchSnapshot();
  });
});
