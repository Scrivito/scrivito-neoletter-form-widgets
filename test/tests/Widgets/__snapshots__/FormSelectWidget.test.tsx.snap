// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FormSelectWidget renders correctly 1`] = `
<div>
  <div>
    <div
      className="select-container mb-3"
    >
      <div
        className="select-title"
      >
        <span
          className="text-super"
        >
           
          Test Select Title
           
        </span>
        <div
          className="mandatory-container"
        >
          <span
            className="text-mandatory"
            onMouseEnter={[Function]}
            onMouseLeave={[Function]}
          >
            *
          </span>
        </div>
        <div
          className="helptext-container"
        >
          <i
            className="bi bi-question-circle"
            onMouseEnter={[Function]}
            onMouseLeave={[Function]}
          />
        </div>
      </div>
      <select
        className="dropdown-select"
        id="staticWidgetID"
        name=""
        required={true}
      >
        <option
          id="empty-option"
          value=""
        />
        <option
          id="0"
          value="Item 1"
        >
          Item 1
        </option>
        <option
          id="1"
          value="Item 2"
        >
          Item 2
        </option>
        <option
          id="2"
          value="Item 3"
        >
          Item 3
        </option>
      </select>
    </div>
  </div>
</div>
`;
